### PR TITLE
Create spec class for HA_RGW

### PR DESCRIPTION
**Pull request by adk3798**
Signed-off-by: Adam King <adking@redhat.com>

NOTE: There are currently failures in the mgr tox on this branch from other commits so the unit test I've added here can't actually be tested until later. It might need some adjustment in the future but I've left it in since I think the general idea is still there.

### CODE REVIEW

**Review by Daniel-Pivonka — APPROVED** [review]
looks good as far as i can tell

**Review by jmolmo — APPROVED** [review]
LGTM

File: src/pybind/mgr/cephadm/tests/test_spec.py [hunk 1]
```diff
@@ -633,3 +634,41 @@ def test_custom_container_spec_config_json():
...
+  virtual_ip_interface: eth0
+  virtual_ip_address: 192.168.20.1/24
+  frontend_port: 8080
+  ha_proxy_port: 1967
+  ha_proxy_stats_enabled: true
+  ha_proxy_stats_user: admin
+  ha_proxy_stats_password: admin
+  ha_proxy_enable_prometheus_exporter: true
+  ha_proxy_monitor_uri: /haproxy_health
+  keepalived_user: admin
```
Comment (jmolmo) [inline src/pybind/mgr/cephadm/tests/test_spec.py]: Remove this line, Not needed.

File: src/pybind/mgr/cephadm/tests/test_spec.py [hunk 2]
```diff
@@ -633,3 +634,41 @@ def test_custom_container_spec_config_json():
...
+    assert spec.virtual_ip_interface == "eth0"
+    assert spec.virtual_ip_address == "192.168.20.1/24"
+    assert spec.frontend_port == 8080
+    assert spec.ha_proxy_port == 1967
+    assert spec.ha_proxy_stats_enabled == True
+    assert spec.ha_proxy_stats_user == "admin"
+    assert spec.ha_proxy_stats_password == "admin"
+    assert spec.ha_proxy_enable_prometheus_exporter == True
+    assert spec.ha_proxy_monitor_uri == "/haproxy_health"
+    assert spec.keepalived_user == "admin"
```
Comment (jmolmo) [inline src/pybind/mgr/cephadm/tests/test_spec.py]: remove

File: src/python-common/ceph/deployment/service_spec.py [hunk 3]
```diff
@@ -779,6 +780,77 @@ def __init__(self,
...
+                 virtual_ip_interfaces: Optional[str] = None,
+                 virtual_ip_address: Optional[str] = None,
+                 frontend_port: Optional[int] = None,
+                 ha_proxy_port: Optional[int] = None,
+                 ha_proxy_stats_enabled: Optional[bool] = None,
+                 ha_proxy_stats_user: Optional[str] = None,
+                 ha_proxy_stats_password: Optional[str] = None,
+                 ha_proxy_enable_prometheus_exporter: Optional[bool] = None,
+                 ha_proxy_monitor_uri: Optional[str] = None,
+                 keepalived_user: Optional[str] = None,
```
Comment (jmolmo) [inline src/python-common/ceph/deployment/service_spec.py]: remove

File: src/python-common/ceph/deployment/service_spec.py [hunk 4]
```diff
@@ -779,6 +780,77 @@ def __init__(self,
...
+        self.virtual_ip_interfaces = virtual_ip_interfaces
+        self.virtual_ip_address = virtual_ip_address
+        self.frontend_port = frontend_port
+        self.ha_proxy_port = ha_proxy_port
+        self.ha_proxy_stats_enabled = ha_proxy_stats_enabled
+        self.ha_proxy_stats_user = ha_proxy_stats_user
+        self.ha_proxy_stats_password = ha_proxy_stats_password
+        self.ha_proxy_enable_prometheus_exporter = ha_proxy_enable_prometheus_exporter
+        self.ha_proxy_monitor_uri = ha_proxy_monitor_uri
+        self.keepalived_user = keepalived_user
```
Comment (jmolmo) [inline src/python-common/ceph/deployment/service_spec.py]: remove

File: src/python-common/ceph/deployment/service_spec.py [hunk 5]
```diff
@@ -779,6 +780,77 @@ def __init__(self,
...
+                'Cannot add HA_RGW: No HA Proxy Stats Password specified')
+        if not self.ha_proxy_enable_prometheus_exporter:
+            raise ServiceSpecValidationError(
+                'Cannot add HA_RGW: HA Proxy Enable Prometheus Exporter option not set')
+        if not self.ha_proxy_monitor_uri:
+            raise ServiceSpecValidationError(
+                'Cannot add HA_RGW: No HA Proxy Monitor Uri specified')
+        if not self.keepalived_user:
+            raise ServiceSpecValidationError(
+                'Cannot add HA_RGW: No Keepalived User specified')
```
Comment (jmolmo) [inline src/python-common/ceph/deployment/service_spec.py]: remove

**Review by jmolmo — CHANGES_REQUESTED** [review]
After taking a look to the keepalived config .. i realize the we do not need the "keepalived_user"

**Review by jmolmo — APPROVED** [review]

**Pull request merged**